File: 3.py
from typing import List

# class Solution:
#     def moveZeroes(self, nums: List[int]) -> None:
#         op = 0
#         i = 0
#         while(op < len(nums)):
#             op += 1
#             if nums[i] == 0:
#                 del nums[i]
#                 nums.append(0)
#             else:
#                 i += 1
#             if op >= len(nums):
#                 break

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        max_length = 0
        start = 0
        end = 0
        dic = {}
        for i, c in enumerate(s):
            end = i
            if c not in dic:
                max_length = max(max_length, end - start + 1)
            else:
                for j in range(start, dic[c]):
                    del dic[s[j]]
                start = dic[c] + 1
            dic[c] = i
        return max_length

if __name__ == '__main__':
    s = 'pwwkew'
    sol = Solution()
    print(sol.lengthOfLongestSubstring(s))

File: 3.cpp
/*
无重复字符的最长子串 滑窗
给定一个字符串 s ，请你找出其中不含有重复字符的 最长子串 的长度。
*/
using namespace std;
#include<vector>
#include<iostream>
#include<algorithm>
#include<numeric>
#include<unordered_map>
#include<math.h>
#include<queue>
#include<stack>

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max_length = 0;
        unordered_map<char, int> map;
        int start = 0, end = 0;
        for(int i = 0; i < s.size(); ++i) {
            end = i;
            auto it = map.find(s[i]);
            if(it == map.end()) {
                max_length = max(end - start + 1, max_length);
            }
            else {
                for(int j = start; j < map[s[i]]; ++j) {
                    map.erase(map.find(s[j]));
                }
                start = map[s[i]] + 1;
            }
            map[s[i]] = i;
        }
        return max_length;
    }
};

int main()
{
	string s = "abcabcbb";
	Solution sol;
	cout << sol.lengthOfLongestSubstring(s) << endl;
	// for(auto num : nums) {
	// 	cout << num << " ";
	// }
	// cout << endl;
	system("pause");
}
